Fix: Nova doesn't work

Yukinobu-Mine opened the pull request:
*Issue #, if available:*
Closes #629 

*Description of changes:*
- If model is Amazon Nova, combine multiple system prompts into one text. https://github.com/aws-samples/bedrock-claude-chat/commit/8084a3072c68d860b78af4fd8770de4e30100245
- If model is Amazon Nova, set the upper limit of `topK` to 128. https://github.com/aws-samples/bedrock-claude-chat/commit/4ec0465f9fe7761a284f514006b33d2f5e70f975
- Remove `title` key from input JSON schema of `AgentTool`. https://github.com/aws-samples/bedrock-claude-chat/commit/dd979e07a82217ca4ca5ff06806d9a593e0fc942
- Optimization for Amazon Nova. https://github.com/aws-samples/bedrock-claude-chat/commit/493b01292918951a8c8651a913386d953af0131c
  - Change the system prompt when doing 'Retrieved Context Citation' with Amazon Nova.
  - If the tool result has more than one element, pass it as single text content formatted as JSON array.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Yukinobu-Mine committed:
If model is Amazon Nova, combine multiple system prompts into one text. #629

Yukinobu-Mine committed:
If model is Amazon Nova, set the upper limit of topK to 128. #629

Yukinobu-Mine committed:
Remove title key from input JSON schema of AgentTool. #629

Yukinobu-Mine committed:
Optimization for Amazon Nova.
- Change the system prompt when doing 'Retrieved Context Citation' with Amazon Nova.
- If the tool result has more than one element, pass it as single text content formatted as JSON array.

Yukinobu-Mine committed:
Fix: mypy errors.

Yukinobu-Mine committed:
Add stack trace to backend error logs.

Yukinobu-Mine committed:
Fix for multimodal tool results for Amazon Nova

statefb commented on backend/app/repositories/models/conversation.py:
```diff
@@ -37,6 +37,11 @@
 from pydantic import BaseModel, Discriminator, Field, JsonValue, field_validator
 
 
+def is_nova_model(model: type_model_name) -> bool:
```
repo/modelsに移動された理由は何かありますか？基本はDDBのモデル定義という体裁なので、bedrock.pyの方のままでも良いかと思いましたが、いかがでしょうか？（ORMの層とドメイン書く層の融合のような形になっていますが。。）

Yukinobu-Mine replied:
@statefb `is_nova_model()` は agent_tool.py から利用するために public にしたのですが、bedrock.py は agent_tool.py を import していて、`is_nova_model()` が bedrock.py にあるままだと循環 import になってしまうため、とりあえずここに移しました。
`app/utils.py` に移すという選択肢も考えていたのですが、そちらの方がいいでしょうか?

Yukinobu-Mine replied:
@statefb `is_nova_model()` を bedrock.py に戻し、循環 import を避けるために `from __future__ import annotations` と `if TYPE_CHECKING` を導入してみました。

statefb commented on backend/app/prompt.py:
このファイルの変更全般的にですが、プロンプトの事情が一見するとわからない（BAD不要とか）のでコメントがあると嬉しいです！

statefb commented on backend/app/agents/tools/agent_tool.py:
```diff
@@ -49,6 +85,16 @@ class InvalidToolError(Exception):
     pass
 
 
+class RemoveTitle(GenerateJsonSchema):
```
「Novaはタイトルがあるとエラーのためカスタムgeneratorを設定」という趣旨のコメントがあると、助かります！

statefb commented on backend/app/bedrock.py:
```diff
@@ -83,7 +80,11 @@ def _prepare_nova_model_params(
     if generation_params and generation_params.top_k is not None:
-        additional_fields["inferenceConfig"]["topK"] = generation_params.top_k
+        top_k = generation_params.top_k
+        if top_k > 128:
```
128超える場合はlog.warning仕込んでも良いかと思いましたが、どうでしょうか？

Yukinobu-Mine committed:
Add comments, and document changes.
- `RemoveTitle` in app/agents/tools/agent_tool.py
- `_prepare_nova_model_params()` in app/bedrock.py
- `build_rag_prompt()` and `get_prompt_to_cite_tool_results()` in app/prompt.py
- `BaseTool` -> `AgentTool` in docs/AGENT.md

Yukinobu-Mine committed:
Move `is_nova_model()` back to app/bedrock.py
- To avoid circular imports, add `from __future__ import annotations` and `if TYPE_CHECKING` to app/bedrock.py

Yukinobu-Mine committed:
Update document of Agent functionality
- docs/AGENT.md

statefb commented on backend/app/agents/tools/agent_tool.py:
```diff
@@ -28,18 +30,53 @@ class ToolRunResult(TypedDict):
+    ]
+    from app.bedrock import is_nova_model
+
+    if is_nova_model(model=model):
```
前回のレビューで漏れてすみません！`related_document.to_tool_result_model`にmodel渡して分岐する方が、同等の処理が一箇所にまとまるため可読性が良いとも思いましたが、いかがでしょうか？

Yukinobu-Mine committed:
Change `run_result_to_tool_result_content_model()` to be an instance method of `ToolResultContentModel`
- `agent_tool.run_result_to_tool_result_content_model()` -> `ToolResultContentModel.from_tool_run_result()`

statefb reviewed (approved):
lgtm
